internal/factsengine/gatherers/saplocalhostresolver.go

```diff
@@ -0,0 +1,154 @@
+package gatherers
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+	"regexp"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/afero"
+	"github.com/trento-project/agent/internal/core/sapsystem"
+	"github.com/trento-project/agent/pkg/factsengine/entities"
+	"github.com/trento-project/agent/pkg/utils"
+)
+
+const (
+	SapLocalhostResolverGathererName = "saplocalhost_resolver"
+)
+
+// nolint:gochecknoglobals
+var (
+	subgroupMapping = map[int]string{
+		1: "SID",
+		2: "InstanceName",
+		3: "Hostname",
+	}
+
+	hostnameParsingRegexp = fmt.Sprintf(
+		`(?P<%s>[A-Z0-9]+)_(?P<%s>[A-Z0-9]+)_(?P<%s>[a-z]+)$`,
+		subgroupMapping[1],
+		subgroupMapping[2],
+		subgroupMapping[3],
+	)
+
+	hostnameRegexCompiled = regexp.MustCompile(hostnameParsingRegexp)
+)
+
+// nolint:gochecknoglobals
+var (
+	SapLocalhostResolverHostnameResolutionError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-resolution-error",
+		Message: "error resolving hostname",
+	}
+	SapLocalhostResolverGathererDecodingError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-decoding-error",
+		Message: "error decoding output to FactValue",
+	}
+	SapLocalhostResolverFileSystemError = entities.FactGatheringError{
+		Type:    "saplocalhost_resolver-file-system-error",
+		Message: "error reading the sap profiles file system",
+	}
+)
+
+type SapLocalhostResolverGatherer struct {
+	fs afero.Fs
+	hr utils.HostnameResolver
+}
+
+type ResolvabilityDetails struct {
+	Hostname     string   `json:"hostname"`
+	Addresses    []string `json:"addresses"`
+	InstanceName string   `json:"instance_name"`
+}
+
+func NewDefaultSapLocalhostResolverGatherer() *SapLocalhostResolverGatherer {
+	return NewSapLocalhostResolver(afero.NewOsFs(), utils.Resolver{})
+}
+
+func NewSapLocalhostResolver(fs afero.Fs, hr utils.HostnameResolver) *SapLocalhostResolverGatherer {
+	return &SapLocalhostResolverGatherer{fs: fs, hr: hr}
+}
+
+func (r *SapLocalhostResolverGatherer) Gather(factsRequests []entities.FactRequest) ([]entities.Fact, error) {
+	facts := make([]entities.Fact, 0, len(factsRequests))
```

Comment:
Please use a direct slice creation without make with a fixed capacity, the code looks less readable and I think we don't need such a thing in this context performance wise